plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.testing.AbstractTestQueryFramework;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Collections.synchronizedMap;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Test(singleThreaded = true) // this class is stateful, see fields
+public abstract class BaseJdbcConnectionCreationTest
+        extends AbstractTestQueryFramework
+{
+    protected ConnectionCountingConnectionFactory connectionFactory;
+
+    @BeforeClass
+    public void verifySetup()
+    {
+        // Test expects connectionFactory to be provided with AbstractTestQueryFramework.createQueryRunner implementation
+        requireNonNull(connectionFactory, "connectionFactory is null");
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected void assertJdbcConnections(@Language("SQL") String query, int expectedJdbcConnectionsCount, Optional<String> errorMessage)
+    {
+        int before = connectionFactory.openConnections.get();
+        if (errorMessage.isPresent()) {
+            assertQueryFails(query, errorMessage.get());
+        }
+        else {
+            getQueryRunner().execute(query);
+        }
+        int after = connectionFactory.openConnections.get();
+        assertThat(after - before).isEqualTo(expectedJdbcConnectionsCount);
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected static class ConnectionCountingConnectionFactory
+            implements ConnectionFactory
+    {
+        // Map from connection to a fake exception (holds stacktrace) pointing to the place where the connection was created
+        private final Map<Connection, Exception> connectionCreations = synchronizedMap(new IdentityHashMap<>());
```

Comment:
clever - unfortunately I can't think of better ways to store this information either